commonmark/src/main/java/org/commonmark/internal/inline/Scanner.java

```diff
@@ -0,0 +1,187 @@
+                line = lines.get(lineIndex);
+                lineLength = line.length();
+            } else {
+                line = "";
+                lineLength = 0;
+            }
+            index = 0;
+        }
+    }
+
+    public boolean next(char c) {
+        if (peek() == c) {
+            next();
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    public int matchMultiple(char c) {
+        int count = 0;
+        while (peek() == c) {
+            count++;
+            next();
+        }
+        return count;
+    }
+
+    public int match(CharMatcher matcher) {
+        int count = 0;
+        while (matcher.matches(peek())) {
+            count++;
+            next();
+        }
+        return count;
+    }
+
+    public int whitespace() {
+        int count = 0;
+        while (true) {
+            switch (peek()) {
+                case ' ':
+                case '\t':
+                case '\n':
+                case '\u000B':
+                case '\f':
+                case '\r':
+                    count++;
+                    next();
+                    break;
+                default:
+                    return count;
+            }
+        }
+    }
+
+    public int find(char c) {
+        int count = 0;
+        while (true) {
+            char cur = peek();
+            if (cur == '\0') {
+                return -1;
+            } else if (cur == c) {
+                return count;
+            }
+            count++;
+            next();
+        }
+    }
+
+    public int find(CharMatcher matcher) {
+        int count = 0;
+        while (true) {
+            char c = peek();
+            if (c == '\0') {
+                return -1;
+            } else if (matcher.matches(c)) {
+                return count;
+            }
+            count++;
+            next();
+        }
+    }
+
+    // Don't expose the int index, because it would be good if we could switch input to a List<String> of lines later
+    // instead of one contiguous String.
+    public Position position() {
+        return new Position(lineIndex, index);
+    }
+
+    // For cases where the caller appends the result to a StringBuilder, we could offer another method to avoid some
+    // unnecessary copying.
+    public CharSequence textBetween(Position begin, Position end) {
+        if (begin.lineIndex == end.lineIndex) {
+            // Shortcut for common case of text from a single line
+            return lines.get(begin.lineIndex).subSequence(begin.index, end.index);
+        } else {
+            StringBuilder sb = new StringBuilder();
+
+            CharSequence firstLine = lines.get(begin.lineIndex);
+            sb.append(firstLine.subSequence(begin.index, firstLine.length()));
+            sb.append('\n');
+
+            // Lines between begin and end (we are appending the full line)
+            for (int line = begin.lineIndex + 1; line < end.lineIndex; line++) {
```

Comment:
We do have an attribute in this line called `line` too. We may call here by simple `i,j,....`